### PR TITLE
prov/sockets: Restructured struct sock_ep to support alias ep

### PR DESCRIPTION
**Pull request by shantonu**
Restructured sock_ep structure to support alias ep and share the resources allocated by the base ep.
- Created a new struct <code>sock_ep_attr</code> that to hold ep resources and modified struct <code>sock_ep</code> to include a pointer to <code>sock_ep_attr</code>
- Added flag <code>is_alias</code> in <code>sock_ep</code> to differentiate alias ep
- Added reference counting for alias ep
- Removed a unused case in <code>sock_pep_listener_thread</code>
- Use correct op_flags
- Verified the patch with fabtests and osu/pt2pt

Modified fabtests/simple/rdm_rma_trigger.c with fi_alias_ep (https://github.com/shantonu/fabtests/tree/pr/alias-test) works with this patch. Will be submitted after this patch is merged.
@jithinjosepkl @shefty 
Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

### CODE REVIEW

**Comment by jithinjosepkl** [comment]
Thanks, looks fine to me 👍 . It will be good to get it verified with the modified fabtests example and get it in.

**Comment by shantonu** [comment]
Yes, I am working on the fabtests patch.

File: prov/sockets/src/sock_rma.c [hunk 1]
```diff
@@ -262,7 +266,7 @@ ssize_t sock_ep_rma_writemsg(struct fid_ep *ep, const struct fi_msg_rma *msg,
 
 	SOCK_EP_SET_TX_OP_FLAGS(flags);
 	if (flags & SOCK_USE_OP_FLAGS)
-		flags |= tx_ctx->attr.op_flags;
+		flags |= op_flags;
```
Comment (shantonu) [inline prov/sockets/src/sock_rma.c]: @jithinjosepkl, can you please verify if these changes are alright? For alias ep, we need to use the ep->op_flags but it uses tx_ctx->attr.op_flags.

Comment (jithinjosepkl) [inline prov/sockets/src/sock_rma.c]: We need to set the tx/rx flags in alias and setopt. That will take care of this.

Comment (shefty) [inline prov/sockets/src/sock_rma.c]: > > @@ -262,7 +266,7 @@ ssize_t sock_ep_rma_writemsg(struct fid_ep *ep, const
> struct fi_msg_rma *msg,
> >
> >  	SOCK_EP_SET_TX_OP_FLAGS(flags);
> >  	if (flags & SOCK_USE_OP_FLAGS)
> > -		flags |= tx_ctx->attr.op_flags;
> > +		flags |= op_flags;
> 
> We need to set the tx/rx flags in alias and setopt. That will take care of
> this.

The op_flags for an alias endpoint separate from the op_flags for the base endpoint.  One cannot change the other.

Comment (jithinjosepkl) [inline prov/sockets/src/sock_rma.c]: Oh yes, I thought tx/rx flags were separate for base/alias eps. So, we need to check the ep->op flags in tx/rx operations.

Comment (jithinjosepkl) [inline prov/sockets/src/sock_rma.c]: But, Sean, operations can be initiated from the tx/rx contexts as well. In this case which op_flags shall be used? the one for base EP?

**Comment by shantonu** [comment]
In #1983, I removed op_flags from sock_ep and used a set of op_flags defined in tx_attr and rx_attr. Now alias ep will have a set of tx and rx attributes instead of a single op_flags. I'll update this patch accordingly.

**Comment by shantonu** [comment]
Updated the patch with manpage and provider updates. Verified with Fabtests patch that @jithinjosepkl  mentioned. @shefty, can you please take a look if you have any additional feedback? When this PR is merged, I'll submit fabtests PR that tests this path.

File: prov/sockets/src/sock_ep_msg.c [hunk 2]
```diff
@@ -1093,11 +1091,6 @@ static void *sock_pep_listener_thread(void *data)
 						conn_req->hdr.msg_id);
 
 		switch (conn_req->hdr.type) {
-		case SOCK_CONN_ACK:
-			SOCK_LOG_DBG("Received SOCK_CONN_ACK\n");
-			sock_ep_cm_handle_ack(&pep->cm, &conn_req->hdr);
-			break;
-
```
Comment (shefty) [inline prov/sockets/src/sock_ep_msg.c]: This change is unrelated.  Why was this code removed?  Is this just an invalid message to process at this point?

**Comment by shefty** [comment]
I see the answer to my question in the comment description.  Next time, please separate changes into distinct patches.

File: prov/sockets/src/sock_ep_msg.c (hunk 2 above)
Comment (shantonu) [inline prov/sockets/src/sock_ep_msg.c]: Thanks! I have changed the signature of <code>sock_ep_cm_handle_ack</code> to take <code>struct sock_ep</code> instead of <code>sock_ep->cm</code> and this case wouldn't work with the same signature as it's a passive ep case. That's why I removed the unused case in the same patch. Sorry for the confusion.
